// ripple-lib/dist/npm/api/transaction/trustline.js

'use strict';
var utils = require('./utils');
var validate = utils.common.validate;
var Transaction = utils.common.core.Transaction;

var TrustSetFlags = {
  authorized: { set: 'SetAuth' },
  allowRippling: { set: 'ClearNoRipple', unset: 'NoRipple' },
  frozen: { set: 'SetFreeze', unset: 'ClearFreeze' }
};

function createTrustlineTransaction(account, trustline) {
  validate.address(account);
  validate.trustline(trustline);

  var limit = {
    currency: trustline.currency,
    issuer: trustline.counterparty,
    value: trustline.limit
  };

  var transaction = new Transaction();
  transaction.trustSet(account, limit, trustline.qualityIn, trustline.qualityOut);
  utils.setTransactionBitFlags(transaction, trustline, TrustSetFlags);
  return transaction;
}

function prepareTrustline(account, trustline, instructions, callback) {
  var transaction = createTrustlineTransaction(account, trustline);
  utils.createTxJSON(transaction, this.remote, instructions, callback);
}

module.exports = utils.wrapCatch(prepareTrustline);
