// ripple-lib/dist/npm/api/common/utils.js
'use strict';
var BigNumber = require('bignumber.js');
var core = require('../../core');
var errors = require('./errors');

function dropsToXrp(drops) {
  return new BigNumber(drops).dividedBy(1000000.0).toString();
}

function xrpToDrops(xrp) {
  return new BigNumber(xrp).times(1000000.0).floor().toString();
}

function toRippledAmount(amount) {
  if (amount.currency === 'XRP') {
    return xrpToDrops(amount.value);
  }
  return {
    currency: amount.currency,
    issuer: amount.counterparty ? amount.counterparty : amount.issuer,
    value: amount.value
  };
}

function wrapCatch(asyncFunction) {
  return function () {
    try {
      asyncFunction.apply(this, arguments);
    } catch (error) {
      var callback = arguments[arguments.length - 1];
      callback(error);
    }
  };
}

function composeAsync(wrapper, callback) {
  return function (error, data) {
    if (error) {
      callback(error);
      return;
    }
    var result = undefined;
    try {
      result = wrapper(data);
    } catch (exception) {
      callback(exception);
      return;
    }
    callback(null, result);
  };
}

function convertExceptions(f) {
  return function () {
    try {
      return f.apply(this, arguments);
    } catch (error) {
      throw new errors.ApiError(error.message);
    }
  };
}

module.exports = {
  core: core,
  dropsToXrp: dropsToXrp,
  xrpToDrops: xrpToDrops,
  toRippledAmount: toRippledAmount,
  wrapCatch: wrapCatch,
  composeAsync: composeAsync,
  convertExceptions: convertExceptions
};
