// ripple-lib/dist/npm/api/ledger/balances.js

'use strict';
var _ = require('lodash');
var async = require('async');
var utils = require('./utils');
var getTrustlines = require('./trustlines');
var validate = utils.common.validate;
var composeAsync = utils.common.composeAsync;

function getTrustlineBalanceAmount(trustline) {
  return {
    currency: trustline.specification.currency,
    counterparty: trustline.specification.counterparty,
    value: trustline.state.balance
  };
}

function formatBalances(balances) {
  var xrpBalance = {
    currency: 'XRP',
    value: balances.xrp
  };
  return [xrpBalance].concat(balances.trustlines.map(getTrustlineBalanceAmount));
}

function getBalances(account, options, callback) {
  validate.address(account);
  validate.getBalancesOptions(options);

  var ledgerVersion = options.ledgerVersion || this.remote.getLedgerSequence();
  async.parallel({
    xrp: _.partial(utils.getXRPBalance, this.remote, account, ledgerVersion),
    trustlines: _.partial(getTrustlines.bind(this), account, options)
  }, composeAsync(formatBalances, callback));
}

module.exports = utils.wrapCatch(getBalances);
